libcaf_net/caf/net/web_socket/framing.cpp

```diff
@@ -95,26 +112,58 @@ ptrdiff_t framing::consume(byte_span buffer, byte_span) {
       return -1;
     }
     // Reject any payload that exceeds max_frame_size. This covers assembled
-    //  payloads as well by including payload_buf_.
+    // payloads as well by including payload_buf_.
     if (payload_buf_.size() + hdr.payload_len > max_frame_size) {
       CAF_LOG_DEBUG("fragmented WebSocket payload exceeds maximum size");
       abort_and_shutdown(sec::protocol_error, "fragmented WebSocket payload "
                                               "exceeds maximum size");
       return -1;
     }
   }
-  // Wait for more data if necessary.
+  // Calculate at what point of the received buffer the delta payload begins.
+  auto offset = static_cast<ptrdiff_t>((buffer.size() - delta.size()))
+                - hdr_bytes;
+  // Offset < zero  - the delta buffer contains header bytes.
+  // Delta is empty - we didn't process the whole input last time we got called.
+  if (delta.empty() || offset < 0)
+    offset = 0;
+  // We read frame_size at most. This can leave unprocessed input.
+  auto payload = buffer.subspan(hdr_bytes, std::min(buffer.size() - hdr_bytes,
+                                                    hdr.payload_len));
```

Comment:
The logic here is quite complex. Calculating the offset, treating special cases and so on. I think this really comes down to leaving the header in the buffer in the first place. So maybe it's actually best to address this right away after all.

So here are some thought that come to mind how we could simplify this logic.

If we keep the parsed header as state in `framing`, we would only need to remember whether we are currently parsing a header or a payload. If we parse a header, we just wait until the transport has handed us enough bytes. Then we return `hdr_bytes`. This means the transport will drop the header bytes for us and we no longer need to "work around it". The next time the transport calls `consume`, we know that we are parsing a payload and can simply call `mask_data` on the delta.

This would also allow us to split the consume function into `consume_header` and `consume_payload` and then `consume` would only dispatch based on the current state.

Does that make sense?